scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,752 @@
+    # config-get               leader-set          resource-get        status-set
+    # containeragent           network-get         secret-add          storage-add
+    # credential-get           open-port           secret-get          storage-get
+    # goal-state               opened-ports        secret-grant        storage-list
+    # is-leader                payload-register    secret-ids          unit-get
+    # juju-log                 payload-status-set  secret-info-get
+    _model = f" -m {model}" if model else ""
+    _target = f" -u {target}" if target else ""
+    return run(
+        f"juju exec{_model}{_target} -- {cmd}".split(), capture_output=True
+    ).stdout.decode("utf-8")
+
+
+def get_leader(target: Target, model: Optional[str]):
+    # could also get it from _juju_run('status')...
+    logger.info("getting leader...")
+    return _juju_exec(target, model, "is-leader") == "True"
+
+
+def get_network(target: Target, model: Optional[str], endpoint: str) -> Network:
+    raw = _juju_exec(target, model, f"network-get {endpoint}")
+    jsn = yaml.safe_load(raw)
+
+    bind_addresses = []
+    for raw_bind in jsn["bind-addresses"]:
+
+        addresses = []
+        for raw_adds in raw_bind["addresses"]:
+            addresses.append(
+                Address(
+                    hostname=raw_adds["hostname"],
+                    value=raw_adds["value"],
+                    cidr=raw_adds["cidr"],
+                    address=raw_adds.get("address", ""),
+                )
+            )
+
+        bind_addresses.append(
+            BindAddress(
+                interface_name=raw_bind.get("interface-name", ""), addresses=addresses
+            )
+        )
+    return Network(
+        name=endpoint,
+        bind_addresses=bind_addresses,
+        egress_subnets=jsn.get("egress-subnets", None),
+        ingress_addresses=jsn.get("ingress-addresses", None),
+    )
+
+
+def get_networks(
+    target: Target,
+    model: Optional[str],
+    metadata: Dict,
+    include_dead: bool = False,
+    relations: Tuple[str, ...] = (),
+) -> List[Network]:
+    logger.info("getting networks...")
+    networks = []
+    networks.append(get_network(target, model, "juju-info"))
+
+    endpoints = relations  # only alive relations
+    if include_dead:
+        endpoints = chain(
+            metadata.get("provides", ()),
+            metadata.get("requires", ()),
+            metadata.get("peers", ()),
+        )
+
+    for endpoint in endpoints:
+        logger.debug(f"  getting network for endpoint {endpoint!r}")
+        networks.append(get_network(target, model, endpoint))
+    return networks
+
+
+def get_metadata(target: Target, model: Optional[str]):
+    logger.info("fetching metadata...")
+
+    raw_meta = _juju_ssh(
+        target,
+        f"cat ./agents/unit-{target.normalized}/charm/metadata.yaml",
+        model=model,
+    )
+    return yaml.safe_load(raw_meta)
```

Comment:
You are calling `yaml.safe_load`, but you do not have `pyyaml` defined as a dependency in your _requirements.txt_ file!

Reply:
it's a transitive dependency from ops no?

Reply:
but yea, if ops were to drop it...